src/tasks/worker.ts

```diff
@@ -0,0 +1,276 @@
+		const body: AnyObject = {
+			started_on__time: startedOnTime,
+			ended_on__time: new Date(),
+			status,
+			attempt_count: attemptCount,
+			...(errorMessage != null && { error_message: errorMessage }),
+		};
+
+		// Re-enqueue if the task failed but has retries left, remember that
+		// attemptCount includes the initial attempt while attempt_limit does not
+		if (status === 'failed' && attemptCount < task.attempt_limit) {
+			body.status = 'queued';
+
+			// Schedule next attempt using exponential backoff
+			body.is_scheduled_to_execute_on__time =
+				this.getNextAttemptTime(attemptCount);
+		}
+
+		// Patch current task
+		await this.client.patch({
+			resource: 'task',
+			passthrough: {
+				tx,
+				req: permissions.root,
+			},
+			id: task.id,
+			body,
+		});
+
+		// Create new task with same configuration if previous
+		// iteration completed and has a cron expression
+		if (
+			['failed', 'succeeded'].includes(body.status) &&
+			task.is_scheduled_with__cron_expression != null
+		) {
+			await this.client.post({
+				resource: 'task',
+				passthrough: {
+					tx,
+					req: permissions.root,
+				},
+				options: {
+					returnResource: false,
+				},
+				body: {
+					attempt_limit: task.attempt_limit,
+					is_created_by__actor: task.is_created_by__actor,
+					is_executed_by__handler: task.is_executed_by__handler,
+					is_executed_with__parameter_set: task.is_executed_with__parameter_set,
+					is_scheduled_with__cron_expression:
+						task.is_scheduled_with__cron_expression,
+				},
+			});
+		}
+	}
+
+	// Calculate next attempt time using exponential backoff
+	private getNextAttemptTime(attempt: number): Date | null {
+		const delay = Math.ceil(Math.exp(Math.min(10, attempt)));
+		return new Date(Date.now() + delay);
+	}
+
+	// Poll for tasks to execute
+	private poll(): void {
+		let executed = false;
+		void (async () => {
+			try {
+				if (!this.canExecute()) {
+					return;
+				}
+				const handlerNames = Object.keys(this.handlers);
+				await sbvrUtils.db.transaction(async (tx) => {
+					const result = await tx.executeSql(
+						`SELECT ${selectColumns}
+						FROM task AS t
+						WHERE
+							t."is executed by-handler" IN (${handlerNames.map((_, index) => `$${index + 1}`).join(', ')}) AND
+							t."status" = 'queued' AND
+							t."attempt count" <= t."attempt limit" AND
+							(
+								t."is scheduled to execute on-time" IS NULL OR
+								t."is scheduled to execute on-time" <= CURRENT_TIMESTAMP + $${handlerNames.length + 1} * INTERVAL '1 SECOND'
+							)
+						ORDER BY
+							t."is scheduled to execute on-time" ASC,
+							t."id" ASC
+						LIMIT $${handlerNames.length + 2}
+						FOR UPDATE SKIP LOCKED`,
+						[
+							...handlerNames,
+							Math.ceil(this.interval / 1000),
+							Math.max(this.concurrency - this.executing, 0),
+						],
+					);
+					if (result.rows.length === 0) {
+						return;
+					}
+
+					// Tasks found, execute them in parallel
```

Comment:
What’s the reasoning behind having one global worker that executes several tasks concurrently instead of several concurrent workers that poll and execute one task each?

At the expense of N poll queries instead of just one (where N=number of workers) you get much simpler code that doesn’t need to juggle multiple things.

Reply:
@dfunckt I think you're raising the same point Page did here: https://github.com/balena-io/pinejs/pull/739#discussion_r1644658715

I completely agree, and will look at implementing this once I have some more time